### ISSUE TITLE
C stored procedure arguments get invalidated when called from lua

### ISSUE THREAD

**Issue by gmoshkin** [tarantool/tarantool#7793]
Arguments passed to C stored procedure become corrupted after some box_* operations are performed in the body of the procedure when called from lua. If called via net.box, the error doesn't reproduce.

If this behavior is allowed, it should probably be mentioned in the docs.

## Repro

**proc.c**:
```c
#include <stdio.h>

int
box_update(
    int space_id, int index_id, char *key,
    char *key_end, char *ops, char *ops_end,
    int index_base, void **result
);

void *
box_tuple_format_default();

void *
box_tuple_new(void *format, char *data, char *end);

int proc(void *_, unsigned char *start, unsigned char *end) {
    unsigned char *p;
    printf("before: ");
    for (p = start; p < end; p++) {
        printf("%02x ", *p);
    }
    printf("\n");

    // random sequence of operations
    // others may or may not work
    char buf[] = {'\x90'};
    box_update(272, 0, buf, buf + 1, buf, buf + 1, 0, 0);
    void *fmt = box_tuple_format_default();
    box_tuple_new(fmt, buf, buf + 1);

    printf("after:  ");
    for (p = start; p < end; p++) {
        printf("%02x ", *p);
    }
    printf("\n");
    return 0;
}
```

**repro.lua**:
```lua
box.cfg{}
box.schema.func.create('proc', {language = 'C', if_not_exists = true})
box.func.proc:call{0xcc, 0xcc, 0xcc, 0xcc}
os.exit(0)
```

How to run:
```bash
gcc -shared -o proc.so -fPIC proc.c
tarantool repro.lua
```

## Expected output:
```
before: 94 cc cc cc cc cc cc cc cc
after:  94 cc cc cc cc cc cc cc cc
```
## Actual output:
```
before: 94 cc cc cc cc cc cc cc cc
after:  50 50 50 50 50 50 50 50 50
```

tarantool --version:
```
Tarantool 2.11.0-entrypoint-582-g805cbaa71
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=TRUE
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
```

**Comment by Totktonada** [comment]
On a brief glance it looks like an effect of #4792.

**Comment by mkostoevr** [comment]
The problem was the `fiber_gc`, which cleared the fiber region in some box functions and led to the use-after-free in the stored procedures. The buffer modification you see is an artefact of the freed slab poisoning (debug-only feature).

A smaller `proc.c` for reproduce:

```C
#include <stdio.h>

#include "module.h"

int proc(void *_, unsigned char *start, unsigned char *end) {
        unsigned char *p;
        printf("before: ");
        for (p = start; p < end; p++) {
                printf("%02x ", *p);
        }
        printf("\n");

        /*
         * Call a box function. This calls fiber_gc() which resets the region.
         * The slabs of the region are still available (aren't returned to the
         * slab cache), so we can still read the valid data.
         */
        char buf[] = {'\x90'};
        box_update(272, 0, buf, buf + 1, buf, buf + 1, 0, 0);

        /*
         * Get the region savepoint and truncate it. The truncation frees the
         * last slab held by the region which triggers the slab poisoning, so
         * we have these 0x50s in the buffer now.
         */
        size_t region_svp = box_region_used();
        box_region_truncate(region_svp);

        printf("after:  ");
        for (p = start; p < end; p++) {
                printf("%02x ", *p);
        }
        printf("\n");
        return 0;
}
```

The issue is fixed in by @nshy in the commit 19abfd2a39205de270836adda7f3a733476acaa4.
